addons/crm/models/crm_team_member.py

```diff
@@ -50,163 +68,95 @@ def _constrains_assignment_domain(self):
-        lead_limit = sum(
-            member._get_assignment_quota(work_days=work_days)
-            for member in members
-        )
-
-        # could probably be optimized
-        for member in members:
-            lead_domain = expression.AND([
-                literal_eval(member.assignment_domain or '[]'),
-                ['&', '&', ('user_id', '=', False), ('date_open', '=', False), ('team_id', '=', member.crm_team_id.id)]
-            ])
-
-            leads = self.env["crm.lead"].search(lead_domain, order='probability DESC, id', limit=lead_limit)
-
-            to_assign = member._get_assignment_quota(work_days=work_days)
-            members_data[member.id] = {
-                "team_member": member,
-                "max": member.assignment_max,
-                "to_assign": to_assign,
-                "leads": leads,
-                "assigned": self.env["crm.lead"],
-            }
-            population.append(member.id)
-            weights.append(to_assign)
-
-        leads_done_ids = set()
-        counter = 0
-        # auto-commit except in testing mode
         auto_commit = not getattr(threading.current_thread(), 'testing', False)
+        result_data = {}
         commit_bundle_size = int(self.env['ir.config_parameter'].sudo().get_param('crm.assignment.commit.bundle', 100))
-        while population and any(weights):
-            counter += 1
-            member_id = random.choices(population, weights=weights, k=1)[0]
-            member_index = population.index(member_id)
-            member_data = members_data[member_id]
-
-            lead = next((lead for lead in member_data['leads'] if lead.id not in leads_done_ids), False)
-            if lead:
-                leads_done_ids.add(lead.id)
-                members_data[member_id]["assigned"] += lead
-                weights[member_index] = weights[member_index] - 1
-
-                lead.with_context(mail_auto_subscribe_no_notify=True).convert_opportunity(
-                    lead.partner_id,
-                    user_ids=member_data['team_member'].user_id.ids
-                )
+        counter = 0
+        for team in self.crm_team_id:
+            team_members = self.filtered(lambda member:
+                member.crm_team_id == team and
+                not member.assignment_optout and
+                member.assignment_max > 0 and
+                member._get_assignment_quota() > 0
+            ).sorted("lead_day_count")
+            if not team_members:
+                break
+            quota_per_member = {member: member._get_assignment_quota(force_quota=force_quota) for member in team_members}
+            result_data.update({member: {'assigned': self.env['crm.lead'], "quota": quota} for member, quota in quota_per_member.items()})
+            members_dom = {m: literal_eval(m.assignment_domain or '[]') for m in team_members}
+            leads = self.env["crm.lead"].search([
+                    ('user_id', '=', False),
```

Comment:
I would add a `for update skip locked` in the select, to reduce concurrency issues (i.e. with lead enrichment). Idea: add the feature in orm search, that could be useful to most cron jobs. (We could add it to lead enrichment select too)

Reply:
in september the cron failed 6 times: 4 of the 6 where not concurrent update but cache miss because the process was trying to access data deleted since the initial read. (due to manual merge probably). Those issue should be partially solved with the cron running more often and thus will have a shorter delay between the initial read the the actual access to the data. An exist could also solve the issue but with a cos of one more select per lead. The 2 other were indeed concurrent update (I think not due to enrichment since there is a mutual exclusion based on the time). I prefer to keep the search with the orm for now and use this feature if we still experiencing a lot of concurrent update in the future